internal/bls/bdn/mask.go

```diff
@@ -0,0 +1,221 @@
+	}
+
+	return m, nil
+}
+
+// Mask returns the bitmask as a byte array.
+func (m *Mask) Mask() []byte {
+	clone := make([]byte, len(m.mask))
+	copy(clone, m.mask)
+	return clone
+}
+
+// Len returns the length of the byte array necessary to store the bitmask.
+func (m *Mask) Len() int {
+	return (len(m.publics) + 7) / 8
+}
+
+// SetMask replaces the current mask by the new one if the length matches.
+func (m *Mask) SetMask(mask []byte) error {
+	if m.Len() != len(mask) {
+		return fmt.Errorf("mismatching mask lengths")
+	}
+
+	m.mask = mask
+	return nil
+}
+
+// GetBit returns true if the given bit is set.
+func (m *Mask) GetBit(i int) (bool, error) {
+	if i >= len(m.publics) || i < 0 {
+		return false, errors.New("index out of range")
+	}
+
+	byteIndex := i / 8
+	mask := byte(1) << uint(i&7)
+	return m.mask[byteIndex]&mask != 0, nil
+}
+
+// SetBit turns on or off the bit at the given index.
+func (m *Mask) SetBit(i int, enable bool) error {
+	if i >= len(m.publics) || i < 0 {
+		return errors.New("index out of range")
+	}
+
+	byteIndex := i / 8
+	mask := byte(1) << uint(i&7)
+	if enable {
+		m.mask[byteIndex] |= mask
+	} else {
+		m.mask[byteIndex] &^= mask
+	}
+	return nil
+}
+
+// forEachBitEnabled is a helper to iterate over the bits set to 1 in the mask
+// and to return the result of the callback only if it is positive.
+func (m *Mask) forEachBitEnabled(f func(i, j, n int) int) int {
+	n := 0
+	for i, b := range m.mask {
+		for j := uint(0); j < 8; j++ {
+			mm := byte(1) << (j & 7)
+
+			if b&mm != 0 {
+				if res := f(i, int(j), n); res >= 0 {
+					return res
+				}
+
+				n++
+			}
+		}
+	}
+
+	return -1
+}
+
+// IndexOfNthEnabled returns the index of the nth enabled bit or -1 if out of bounds.
+func (m *Mask) IndexOfNthEnabled(nth int) int {
+	return m.forEachBitEnabled(func(i, j, n int) int {
+		if n == nth {
+			return i*8 + int(j)
+		}
+
+		return -1
+	})
+}
+
+// NthEnabledAtIndex returns the sum of bits set to 1 until the given index. In other
+// words, it returns how many bits are enabled before the given index.
+func (m *Mask) NthEnabledAtIndex(idx int) int {
+	return m.forEachBitEnabled(func(i, j, n int) int {
+		if i*8+int(j) == idx {
+			return n
+		}
+
+		return -1
+	})
+}
+
+// Publics returns a copy of the list of public keys.
+func (m *Mask) Publics() []kyber.Point {
+	pubs := make([]kyber.Point, len(m.publics))
+	copy(pubs, m.publics)
+	return pubs
+}
+
+// Participants returns the list of public keys participating.
+func (m *Mask) Participants() []kyber.Point {
+	pp := []kyber.Point{}
+	for i, p := range m.publics {
+		byteIndex := i / 8
+		mask := byte(1) << uint(i&7)
+		if (m.mask[byteIndex] & mask) != 0 {
+			pp = append(pp, p)
+		}
+	}
+
+	return pp
+}
+
+// CountEnabled returns the number of bit set to 1
+func (m *Mask) CountEnabled() int {
+	count := 0
+	for i := range m.publics {
+		byteIndex := i / 8
+		mask := byte(1) << uint(i&7)
+		if (m.mask[byteIndex] & mask) != 0 {
+			count++
```

Comment:
We could just popcount but doens't matter.

Reply:
I'll make a PR upstream.

Reply:
Hm. Later. I have too many stacked.